data-plane/src/server/data_plane_server.rs

```diff
@@ -54,88 +56,353 @@ where
+                    }
+                }
             }
         });
     }
 }
 
+async fn auth_request_non_http(
+    headers: &[httparse::Header<'_>],
+    e3_client: Arc<E3Client>,
+) -> Result<()> {
+    let api_key = headers
+        .iter()
+        .find(|header| header.name.to_ascii_lowercase() == "api-key")
+        .ok_or(MissingApiKey)?;
+    let header_value = HeaderValue::from_bytes(api_key.value)?;
+    if FeatureContext::get().api_key_auth
+        && auth_request(header_value, CageContext::get()?, e3_client)
+            .await
+            .is_some()
+    {
+        return Err(ApiKeyInvalid);
+    }
+    Ok(())
+}
+
+async fn shutdown_conn<L>(stream: &mut TlsStream<L>)
+where
+    TlsStream<L>: AsyncWriteExt + Unpin,
+{
+    if let Err(e) = stream.shutdown().await {
+        eprintln!("Failed to shutdown data plane connection — {e:?}");
+    }
+}
+
+async fn process_stream<L>(stream: &mut TlsStream<L>) -> Result<Vec<u8>>
+where
+    TlsStream<L>: AsyncReadExt + Unpin,
+{
+    let mut buffer = Vec::new();
+    loop {
+        let mut request_chunk = [0; 1024];
+        match stream.read(&mut request_chunk).await {
+            Ok(read_bytes) => {
+                if read_bytes == 0 {
+                    break;
+                }
+
+                buffer.extend_from_slice(&request_chunk[..read_bytes]);
+
+                if read_bytes < request_chunk.len() {
+                    break;
+                }
+            }
+            Err(e) => {
+                eprintln!("Error reading from stream: {}", e);
+                return Err(crate::error::Error::Io(e));
+            }
+        }
+    }
+    Ok(buffer)
+}
+
+async fn pipe_to_customer_process<L>(
+    stream: &mut TlsStream<L>,
+    buffer: &[u8],
+    port: u16,
+) -> Result<()>
+where
+    TlsStream<L>: AsyncReadExt + Unpin + AsyncWrite,
+{
+    let mut customer_stream = TcpStream::connect(("127.0.0.1", port)).await.unwrap();
+    customer_stream.write_all(buffer).await.unwrap();
+    pipe_streams(stream, customer_stream).await?;
+    Ok(())
+}
+
+fn build_http_request(
+    request: httparse::Request<'_, '_>,
+    body_buffer: &[u8],
+    port: u16,
+) -> Result<Request<Body>> {
+    let uri = format!("http://127.0.0.1:{}{}", port, request.path.unwrap());
+    let mut header_map = HeaderMap::new();
+    for header in request.headers {
+        header_map.insert(
+            HeaderName::from_str(header.name)?,
+            HeaderValue::from_bytes(header.value)?,
+        );
+    }
+
+    let mut req = Request::builder()
+        .uri(uri)
+        .method(request.method.unwrap())
+        .body(Body::from(
+            String::from_utf8(body_buffer.to_vec()).expect("Found invalid UTF-8"),
+        ))?;
+    *req.headers_mut() = header_map;
+    Ok(req)
+}
+
+async fn response_to_bytes(response: Response<Body>) -> Vec<u8> {
+    let mut bytes = Vec::new();
+
+    // TODO: handle other versions
+    let status_line = format!(
+        "HTTP/{} {} {}\r\n",
+        "1.1",
```

Comment:
The response type of hyper exposes the HTTP version

Reply:
Ahh nice thanks will use that